### PR TITLE
Added non uniform scaling support for bounds control

### PR DESCRIPTION
**Pull request by thalbern**
## Overview
- Added precise scaling mode to bounds control scale handles. 
- Added test to verify behavior for near interaction with control.

Note: there's another branch where I integrated the constraint manager into BC which will allow to combine non uniform scaling with scale axis locking. 

## Changes
- Fixes: #6734 

nonuniformscale

## Verification
- added new test to verify proper scaling behavior as well as position / center modification when scaling is performed 
- manual tests on example scenes

### CODE REVIEW

**Comment by thalbern** [comment]
by accident i created this in main repo and not in my fork 🤦

File: Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControlTypes.cs [hunk 1]
```diff
@@ -107,4 +107,19 @@ internal enum CardinalAxisType
...
+    public enum HandleScaleMode
+    {
+        /// <summary>
+        /// Control will be scaled uniformly.
+        /// </summary>
+        Uniform,
+        /// <summary>
+        /// Precise scaling scales non uniformly according to movement in 3d space.
+        /// </summary>
+        Precise
```
Comment (julenka) [inline Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControlTypes.cs:123]: I am  not that non-uniform is the same as precise. Should we just call it non-uniform? Precise could be a separate mode...

**Comment by julenka** [comment]
From UX POV, it seems like if we are in a non-uniform scale mode, the affordances should look different: the cursor in far mode should look different, and the corners should look different. @cre8ivepark  should we perhaps collaborate with Bernadette on some designs?

I don't like this idea, but maybe we can use something like the move cursor:

image

And then maybe something with arrows for the corner (I also don't like it but it's a first stab so that others can suggest other ideas):

image

File: Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControlTypes.cs (hunk 1 above)
Comment (thalbern) [reply Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControlTypes.cs:123]: i wasn't sure about the naming - i can change it to non-uniform. by saying precise is a different mode are you referring to the precision tools from finn?

**Comment by thalbern** [comment]
> From UX POV, it seems like if we are in a non-uniform scale mode, the affordances should look different: the cursor in far mode should look different, and the corners should look different. @cre8ivepark Yoon Park FTE should we perhaps collaborate with Bernadette on some designs?
> 
> I don't like this idea, but maybe we can use something like the move cursor:
> 
> image
> 
> And then maybe something with arrows for the corner (I also don't like it but it's a first stab so that others can suggest other ideas):
> 
> image

Sounds good! we can do that with the cursor context component!  I'll create a task for that - we can do that after graduation as well as it shouldn't affect the control itself :)

File: Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControlTypes.cs (hunk 1 above)
Comment (thalbern) [reply Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControlTypes.cs]: changed to non uniform

File: Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs [hunk 2]
```diff
@@ -268,6 +269,41 @@ public IEnumerator ScaleViaNearInteraction()
...
+        /// <summary>
+        /// Uses near interaction to scale the bounds control by directly grabbing corner - precise scaling 
+        /// </summary>
+        [UnityTest]
+        public IEnumerator ScaleNonUniform()
+        {
+            BoundsControl boundsControl = InstantiateSceneAndDefaultBoundsControl();
+            boundsControl.ScaleHandlesConfig.ScaleBehavior = HandleScaleMode.NonUniform;
+            yield return VerifyInitialBoundsCorrect(boundsControl);
+            var inputSimulationService = PlayModeTestUtilities.GetInputSimulationService();
```
Comment (vaoliva) [inline Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs:281]: Is inputSimulationService being used somewhere?

File: Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs [hunk 3]
```diff
@@ -268,6 +269,41 @@ public IEnumerator ScaleViaNearInteraction()
...
+            yield return null;
+
+            var endBounds = boundsControl.GetComponent<BoxCollider>().bounds;
+            Vector3 expectedCenter = new Vector3(0.05f, 0.05f, 1.5f);
+            Vector3 expectedSize = Vector3.one * .6f;
+            expectedSize.z = 0.5f;
+            TestUtilities.AssertAboutEqual(endBounds.center, expectedCenter, "endBounds incorrect center");
+            TestUtilities.AssertAboutEqual(endBounds.size, expectedSize, "endBounds incorrect size");
+            
+            GameObject.Destroy(boundsControl.gameObject);
```
Comment (vaoliva) [inline Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs:302]: Out of curiosity, the scene game objects are being destroyed on the test TearDown inside PlayModeTestUtilities.TearDown(). Is there a special need for calling this here?

File: Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControl.cs [hunk 4]
```diff
@@ -1024,24 +1023,38 @@ private void TransformTarget(HandleType transformType)
...
+                        float scaleFactorUniform = 1 + (currentDist - initialDist) / initialDist;
+                        scaleFactor = new Vector3(scaleFactorUniform, scaleFactorUniform, scaleFactorUniform);
+                    }
+                    else // non-uniform scaling
+                    {
+                        // get diff from center point of box
+                        Vector3 initialDist = (initialGrabPoint - oppositeCorner);
+                        Vector3 currentDist = (currentGrabPoint - oppositeCorner);
+                        Vector3 grabDiff = (currentDist - initialDist);
+                        scaleFactor = Vector3.one + grabDiff.Div(initialDist);
```
Comment (vaoliva) [inline Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControl.cs:1041]: Can the user achieve this kind of interaction with the changes?

bounds_control_scale

Would be nice if we could have the option of specific handles per axis. Usefull when content layout needs to be maintained by only scaling the content mask object per axis.

File: Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs (hunk 2 above)
Comment (thalbern) [reply Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs:281]: actually not anymore - i changed usage from playmodetestutilities to test hand. can be removed!

File: Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControl.cs (hunk 4 above)
Comment (thalbern) [reply Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControl.cs:1041]: you can do that if you combine the non uniform scaling with the constraint manager - however the scale constraints are atm fixed on the corners - if you would want them to appear on the sides you would have to create your own version of scalehandles

File: Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs (hunk 3 above)
Comment (thalbern) [reply Assets/MRTK/Tests/PlayModeTests/Experimental/BoundsControlTests.cs:302]: no we don't need it. i'll remove it

File: Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControl.cs (hunk 4 above)
Comment (thalbern) [reply Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/BoundsControl.cs:1041]: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/8302

**Pull request merged**